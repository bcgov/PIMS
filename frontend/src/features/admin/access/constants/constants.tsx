import { ColumnWithProps } from 'components/Table';
import { AccessStatusDisplayMapper } from 'constants/accessStatus';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IAccessRequestModel } from '../interfaces';

export const columnDefinitions: ColumnWithProps<IAccessRequestModel>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IAccessRequestModel>) => {
      return (
        <Link to={`/admin/user/${props.row.original.userId}`}>{props.row.original.username}</Link>
      );
    },
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Status',
    accessor: 'status',
    align: 'left',
    clickable: true,
    width: 100,
    Cell: (props: CellProps<IAccessRequestModel>) =>
      AccessStatusDisplayMapper[props.row.original.status],
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Role',
    accessor: 'role',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
