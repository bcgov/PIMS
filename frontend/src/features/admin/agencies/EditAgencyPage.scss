@import '../../../variables';
.navBar {
  font-weight: 700;
  text-align: left;
}

.required {
  color: red;
  font-weight: 700;
}

.agency-edit-form-container {
  justify-content: center;
  margin-top: 20px;
  margin-bottom: $footer-height;
  .agencyInfo {
    .label,
    .form-label {
      float: left;
    }
    .buttons {
      .cancelSave {
        margin-top: 100px;
      }
    }
    .checkboxes {
      margin-top: 10px;
      float: left;
      margin-bottom: 10px;
    }
  }
}
