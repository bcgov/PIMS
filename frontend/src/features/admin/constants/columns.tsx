import { ColumnWithProps } from 'components/Table';
import { IAgency, IAgencyRecord } from 'interfaces';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

export const columnDefinitions: ColumnWithProps<IAgencyRecord>[] = [
  {
    Header: 'Agency name',
    accessor: 'name',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Short name',
    accessor: 'code',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Description',
    accessor: 'description',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Parent Agency',
    accessor: 'parent',
    align: 'left',
    clickable: true,
    Cell: (props: CellProps<IAgency>) => {
      return (
        <Link to={`/admin/agency/${props.row.original.parentId}`}>{props.row.original.parent}</Link>
      );
    },
  },
];

export const adminAreasColumnDefinistions: ColumnWithProps<any>[] = [
  {
    Header: 'Area name',
    accessor: 'name',
    align: 'left',
    clickable: true,
  },
];
