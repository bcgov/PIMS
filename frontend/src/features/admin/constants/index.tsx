import * as React from 'react';
import { ColumnWithProps } from 'components/Table';
import { CellProps } from 'react-table';
import { Link } from 'react-router-dom';
import { IAgency, IAgencyRecord } from 'interfaces';

export const columnDefinitions: ColumnWithProps<IAgencyRecord>[] = [
  {
    Header: 'Agency name',
    accessor: 'name',
    align: 'left',
    sortable: true,
    clickable: true,
  },
  {
    Header: 'Short name',
    accessor: 'code',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Description',
    accessor: 'description',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Parent Agency',
    accessor: 'parent',
    align: 'left',
    clickable: true,
    Cell: (props: CellProps<IAgency>) => {
      return (
        <Link to={`/admin/agency/${props.row.original.parentId}`}>{props.row.original.parent}</Link>
      );
    },
    sortable: true,
  },
];
