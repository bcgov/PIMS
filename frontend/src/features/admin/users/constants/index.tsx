import { ColumnWithProps } from 'components/Table';
import * as React from 'react';
import { FaCheck, FaTimes } from 'react-icons/fa';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IUserRecord } from '../interfaces/IUserRecord';

export const columnDefinitions: ColumnWithProps<IUserRecord>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IUserRecord>) => {
      return <Link to={`/admin/user/${props.row.original.id}`}>{props.row.original.username}</Link>;
    },
    sortable: true,
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Active',
    accessor: 'isDisabled',
    align: 'left',
    clickable: true,
    sortable: true,
    width: 100,
    Cell: (props: CellProps<IUserRecord>) =>
      props.row.original.isDisabled ? <FaTimes /> : <FaCheck />,
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    sortable: true,
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Roles',
    accessor: 'roles',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Last Login',
    accessor: 'lastLogin',
    align: 'left',
    clickable: true,
    minWidth: 100,
  },
  {
    Header: 'Created On',
    accessor: 'createdOn',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 100,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
