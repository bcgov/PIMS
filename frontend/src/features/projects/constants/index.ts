export * from './AgencyResponses';
export * from './DisposalWorkflows';
export * from './DisposeWorkflowStatus';
export * from './ProjectWorkflowComponents';
export * from './ReviewWorkflowStatus';
export * from './SPPApprovalTabs';
