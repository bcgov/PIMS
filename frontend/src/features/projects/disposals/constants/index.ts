export * from './defaultFilter';
export * from './defaultProjectForm';
