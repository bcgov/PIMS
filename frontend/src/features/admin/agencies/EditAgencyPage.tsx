import React, { useEffect, useState } from 'react';
import { Navbar, Container, Row, ButtonToolbar, Button } from 'react-bootstrap';
import { Check, Form, Input, Select, SelectOption } from '../../../components/common/form';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from 'reducers/rootReducer';
import { IAgencyDetail } from 'interfaces';
import { Formik } from 'formik';
import * as API from 'constants/API';
import './EditAgencyPage.scss';
import { useHistory } from 'react-router-dom';
import useCodeLookups from 'hooks/useLookupCodes';
import { ILookupCode } from 'actions/lookupActions';
import {
  createAgency,
  deleteAgency,
  fetchAgencyDetail,
  getUpdateAgencyAction,
} from 'actionCreators/agencyActionCreator';
import { FaArrowAltCircleLeft } from 'react-icons/fa';
import GenericModal from 'components/common/GenericModal';

interface IEditAgencyPageProps {
  /** id prop to identify which agency to edit */
  id: number;
  match?: any;
}

/** This page is used to either add a new agency or edit the and agency's details */
const EditAgencyPage = (props: IEditAgencyPageProps) => {
  const agencyId = props?.match?.params?.id || props.id;
  const history = useHistory();
  const dispatch = useDispatch();
  const newAgency = history.location.pathname.includes('/new');
  const [showDelete, setShowDelete] = useState(false);
  useEffect(() => {
    if (!newAgency) {
      dispatch(fetchAgencyDetail({ id: agencyId }));
    }
  }, [dispatch, agencyId, newAgency]);

  const { getByType } = useCodeLookups();
  const agencies = getByType(API.AGENCY_CODE_SET_NAME);

  const agency = useSelector<RootState, IAgencyDetail>(
    state => state.GET_AGENCY_DETAIL as IAgencyDetail,
  );
  const mapLookupCode = (code: ILookupCode): SelectOption => ({
    label: code.name,
    value: code.id.toString(),
    selected: !!agency.parentId,
  });
  //
  const selectAgencies = agencies.map(c => mapLookupCode(c));

  const checkAgencies = (
    <Select
      label="Parent Agency - If applicable"
      field="parentId"
      data-testid="agency"
      options={selectAgencies}
      placeholder={newAgency ? 'Please select if applicable' : undefined}
    />
  );

  const goBack = () => {
    history.push('/admin/agencies');
  };

  const newValues: any = {
    code: '',
    name: '',
    description: '',
    isDisabled: false,
    sendEmail: true,
    email: '',
    rowVersion: '',
  };

  const initialValues: IAgencyDetail = {
    ...agency,
  };

  return (
    <div>
      {showDelete && <DeleteModal {...{ showDelete, setShowDelete, history, dispatch, agency }} />}
      <Navbar className="navBar" expand="sm" variant="light" bg="light">
        <Navbar.Brand href="#">
          {' '}
          <FaArrowAltCircleLeft onClick={goBack} size={20} /> Agency Information
        </Navbar.Brand>
      </Navbar>
      <Container fluid={true}>
        <Row className="agency-edit-form-container">
          <Formik
            enableReinitialize
            initialValues={newAgency ? newValues : initialValues}
            onSubmit={async (values, { setSubmitting, setStatus }) => {
              try {
                if (!newAgency) {
                  dispatch(
                    getUpdateAgencyAction(
                      { id: agencyId },
                      {
                        id: agency.id,
                        name: values.name,
                        code: values.code,
                        email: values.email,
                        isDisabled: values.isDisabled,
                        sendEmail: values.sendEmail,
                        parentId: Number(values.parentId),
                        description: values.description,
                        rowVersion: values.rowVersion,
                      },
                    ),
                  );
                } else {
                  const data = await createAgency({
                    name: values.name,
                    code: values.code,
                    email: values.email,
                    isDisabled: values.isDisabled,
                    sendEmail: values.sendEmail,
                    parentId: Number(values.parentId),
                    description: values.description,
                  })(dispatch);
                  history.replace(`/admin/agency/${data.id}`);
                }
              } catch (error) {
                const msg: string =
                  error?.response?.data?.error ?? 'Error saving property data, please try again.';
                setStatus({ msg });
              } finally {
                setSubmitting(false);
              }
            }}
          >
            {props => (
              <Form className="agencyInfo">
                <Input label="Agency" required field="name" value={props.values.name} type="text" />
                <Input
                  label="Short Name (Code)"
                  field="code"
                  value={props.values.code}
                  type="text"
                />
                {checkAgencies}
                <Input label="Description" field="description" type="text" />
                <Input label="Agency e-mail address" field="email" type="text" />

                <Form.Group className="checkboxes">
                  <Check
                    toolTip="Toggle to change Agency status and click Save"
                    toolTipId="is-disabled-tooltip"
                    field="isDisabled"
                    label="Is Disabled?"
                  />
                  <Check
                    toolTip="Toggle to change whether to send an email and click Save"
                    toolTipId="email-tooltip"
                    field="sendEmail"
                    label="Send email?"
                  />
                </Form.Group>

                <hr></hr>
                <Row className="buttons">
                  <ButtonToolbar className="cancelSave">
                    {!newAgency ? (
                      <Button
                        className="bg-danger mr-5"
                        type="button"
                        onClick={() => setShowDelete(true)}
                      >
                        Delete Agency
                      </Button>
                    ) : (
                      <Button className="bg-danger mr-5" type="button" onClick={() => goBack()}>
                        Cancel
                      </Button>
                    )}

                    <Button className="mr-5" type="submit">
                      {newAgency ? 'Save' : 'Save Changes'}
                    </Button>
                  </ButtonToolbar>
                </Row>
              </Form>
            )}
          </Formik>
        </Row>
      </Container>
    </div>
  );
};

export default EditAgencyPage;

const DeleteModal = ({ showDelete, setShowDelete, history, dispatch, agency }: any) => (
  <GenericModal
    message="Are you sure you want to permanently delete the agency?"
    cancelButtonText="Cancel"
    okButtonText="Delete"
    display={showDelete}
    handleOk={() => {
      dispatch(deleteAgency(agency)).then(() => {
        history.push('/admin/agencies');
      });
    }}
    handleCancel={() => {
      setShowDelete(false);
    }}
  />
);
