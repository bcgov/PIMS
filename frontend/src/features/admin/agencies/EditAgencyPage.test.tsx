import EditAgencyPage from './EditAgencyPage';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { ILookupCode } from 'actions/lookupActions';
import * as API from 'constants/API';
import { Provider } from 'react-redux';
import * as reducerTypes from 'constants/reducerTypes';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { render, cleanup, act } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { ToastContainer } from 'react-toastify';
import pretty from 'pretty';

const mockStore = configureMockStore([thunk]);
const history = createMemoryHistory();

const lCodes = {
  lookupCodes: [
    { name: 'Test Agency', id: '111', isDisabled: false, type: API.AGENCY_CODE_SET_NAME },
  ] as ILookupCode[],
};

const selectedAgency = {
  code: 'TEST',
  id: 111,
  email: 'test@email.com',
  isDisabled: false,
  sendEmail: true,
  type: 'Agency',
  name: 'Test Agency',
};

const store = mockStore({
  [reducerTypes.GET_AGENCY_DETAIL]: selectedAgency,
  [reducerTypes.LOOKUP_CODE]: lCodes,
});

const mockAxios = new MockAdapter(axios);

const renderEditAgencyPage = () =>
  render(
    <Provider store={store}>
      <Router history={history}>
        <ToastContainer
          autoClose={5000}
          hideProgressBar
          newestOnTop={false}
          closeOnClick={false}
          rtl={false}
          pauseOnFocusLoss={false}
        />
        <EditAgencyPage id={111} />,
      </Router>
    </Provider>,
  );

describe('Edit agency page', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });
  beforeEach(() => {
    mockAxios.onAny().reply(200, {});
  });
  it('EditAgencyPage renders', () => {
    const { container } = render(
      <Provider store={store}>
        <Router history={history}>
          <EditAgencyPage id={111} />,
        </Router>
      </Provider>,
    );
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  describe('appropriate fields are autofilled', () => {
    it('autofills  email, name, send email, and code', () => {
      const { getByLabelText } = renderEditAgencyPage();
      expect(getByLabelText(/agency e-mail address/i).getAttribute('value')).toEqual(
        'test@email.com',
      );
      expect(getByLabelText('Agency').getAttribute('value')).toEqual('Test Agency');
      expect(getByLabelText('Short Name (Code)').getAttribute('value')).toEqual('TEST');
      expect(getByLabelText(/email notifications?/i).getAttribute('value')).toEqual('true');
    });
  });

  describe('when the agency edit form is submitted', () => {
    it('displays a loading toast', async done => {
      const { getByText, findByText } = renderEditAgencyPage();
      const saveButton = getByText(/save/i);
      act(() => {
        saveButton.click();
      });
      await findByText('Updating Agency...');
      done();
    });

    it('displays a success toast if the request passes', async done => {
      const { getByText, findByText } = renderEditAgencyPage();
      const saveButton = getByText(/save/i);
      act(() => {
        saveButton.click();
      });
      await findByText('Agency updated');
      done();
    });

    it('displays a error toast if the request fails', async done => {
      const { getByText, findByText } = renderEditAgencyPage();
      const saveButton = getByText(/save/i);
      mockAxios.reset();
      mockAxios.onAny().reply(500, {});
      act(() => {
        saveButton.click();
      });
      await findByText('Failed to update Agency');
      done();
    });
  });
});
