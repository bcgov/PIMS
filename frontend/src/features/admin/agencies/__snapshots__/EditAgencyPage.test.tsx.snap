// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit agency page EditAgencyPage renders 1`] = `
<div>
  <nav
    class="navBar navbar navbar-expand-sm navbar-light bg-light"
  >
    <span
      class="navbar-brand"
    >
       
      <svg
        fill="currentColor"
        height="20"
        size="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 504C119 504 8 393 8 256S119 8 256 8s248 111 248 248-111 248-248 248zm116-292H256v-70.9c0-10.7-13-16.1-20.5-8.5L121.2 247.5c-4.7 4.7-4.7 12.2 0 16.9l114.3 114.9c7.6 7.6 20.5 2.2 20.5-8.5V300h116c6.6 0 12-5.4 12-12v-64c0-6.6-5.4-12-12-12z"
        />
      </svg>
    </span>
    <h4>
      Agency Information
    </h4>
  </nav>
  <div
    class="container-fluid"
  >
    <div
      class="agency-edit-form-container row"
    >
      <form
        class="agencyInfo"
        novalidate=""
      >
        <div
          class="required form-group"
        >
          <label
            class="form-label"
            for="input-name"
          >
            Agency
          </label>
          <input
            class="form-control"
            id="input-name"
            name="name"
            required=""
            type="text"
            value="Test Agency"
          />
        </div>
        <div
          class="required form-group"
        >
          <label
            class="form-label"
            for="input-code"
          >
            Short Name (Code)
          </label>
          <input
            class="form-control"
            id="input-code"
            name="code"
            required=""
            type="text"
            value="TEST"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="input-parentId"
          >
            Parent Agency - If applicable
          </label>
          <select
            class="form-select form-control"
            disabled=""
            id="input-parentId"
            name="parentId"
          >
            <option
              value=""
            >
              No parent
            </option>
            <option
              class="option"
              value="111"
            >
              Test Agency
            </option>
          </select>
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="input-description"
          >
            Description
          </label>
          <input
            class="form-control"
            id="input-description"
            name="description"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="input-email"
          >
            Agency e-mail address
          </label>
          <input
            class="form-control"
            id="input-email"
            name="email"
            type="text"
            value="test@email.com"
          />
        </div>
        <div
          class="checkboxes form-group"
        >
          <div
            class="form-group"
          >
            <div
              class="check-field"
            >
              <label
                class="form-label"
                for="input-isDisabled"
              >
                Disable Agency?
              </label>
              <div
                class="form-check"
              >
                <input
                  class="form-check-input position-static"
                  id="input-isDisabled"
                  name="isDisabled"
                  type="checkbox"
                  value="false"
                />
              </div>
            </div>
          </div>
          <div
            class="form-group"
          >
            <div
              class="check-field"
            >
              <label
                class="form-label"
                for="input-sendEmail"
              >
                Email Notifications?
              </label>
              <div
                class="form-check"
              >
                <input
                  checked=""
                  class="form-check-input position-static"
                  id="input-sendEmail"
                  name="sendEmail"
                  type="checkbox"
                  value="true"
                />
              </div>
            </div>
          </div>
        </div>
        <hr />
        <div
          class="buttons row"
        >
          <div
            class="cancelSave btn-toolbar"
            role="toolbar"
          >
            <button
              class="bg-danger mr-5 btn btn-primary"
              type="button"
            >
              Delete Agency
            </button>
            <button
              class="mr-5 btn btn-primary"
              type="submit"
            >
              Save Changes
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
